_posts/2024-01-26-cats-effect-parallelism.md

```diff
@@ -0,0 +1,789 @@
+
+Please, have a look at the [GitHub docs](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) to do it.
+
+## 4. Planning
+What are we going to do and how exactly? To answer this questions we must investigate existing GitHub API so that we know what kind of data is exposed and how.
+
+Some of the interesting questions can be:
+- How can I fetch basic data about organization?
+- How do I find out how many public projects are owned by the organization?
+- What is the number of contributors for each project owned by organization?
+- ...
+
+## 5. Quickstart
+
+Let's just create a `Main.scala` and run a basic HTTP server with health check endpoint to ensure that everything is fine.
+ 
+```scala
+package com.rockthejvm
+
+import cats.effect.{IO, IOApp}
+import org.http4s.dsl.io.*
+import org.http4s.ember.server.EmberServerBuilder
+import org.http4s.HttpRoutes
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import org.typelevel.log4cats.syntax.LoggerInterpolator
+import com.comcast.ip4s.*
+
+object Main extends IOApp.Simple {
+
+  given logger: Logger[IO] = Slf4jLogger.getLogger[IO]
+
+  override val run: IO[Unit] =
+    (for {
+      _ <- info"starting server".toResource
+      _ <- EmberServerBuilder
+        .default[IO]
+        .withHost(host"localhost")
+        .withPort(port"9000")
+        .withHttpApp(routes.orNotFound)
+        .build
+    } yield ()).useForever
+
+  def routes: HttpRoutes[IO] =
+    HttpRoutes.of[IO] { case GET -> Root =>
+      Ok("hi :)")
+    }
+}
+```
+
+To start the app we need to execute `sbt` and `run` commands:
+
+- `sbt`
+- `run`
+
+and then test our server by curling it:
+
+`curl localhost:9000`
+
+To illustrate all that:
+![alt ""](../images/github-contributors-aggregator/quickstart.png)
+
+Great! Now it's time to move on to some domain modeling.
+
+## 6. Domain Modeling
+
+It turns out that GitHub API responses support JSON format, it means that we'll need to define custom JSON deserializers for our domain models.
+
+Since we're working on "GitHub Organization Contributors Aggregator" we'd probably need to think in terms of following nouns:
+- Organization
+- Project
+- Contributor
+- ...
+
+Also, it's important to keep in mind what is returned by each REST API request so that our deserializers do not fail unexpectedly.
+
+Let's start by Organization. Organization is a group of developers that develops multiple open source projects, however the only thing we're interested is the amount of public repositories it owns.
+
+Turns out that if we send a basic request to ```"https://api.github.com/orgs/$orgName"``` we will get a JSON response back, in which we could find the amount of public repositories.
```

Comment:
We could, but I avoided that because from the response JSON we only care about the "amount of public repositories", thus:
```scala
opaque type PublicRepos = Int
```

I think adding a case class for Organization won't give us any benefits anyways, or am I missing your point?

Reply:
I just included `Organization` in those list of domain objects because that's what we would think about it for the first time. So this list can be treated as a "probable domain objects" but it turns out that we don't really need to model `Organization` in the end, we just need the amount of public repositories from that response.